src/ocf_seq_cutoff.c

```diff
@@ -146,55 +205,137 @@ bool ocf_core_seq_cutoff_check(ocf_core_t core, struct ocf_request *req)
-		stream->bytes += req->byte_length;
+		stream->last = addr + len;
+		stream->bytes += len;
+		stream->req_count++;
 		if (!can_update) {
-			ocf_rb_tree_remove(&core->seq_cutoff->tree, node);
-			ocf_rb_tree_insert(&core->seq_cutoff->tree, node);
+			ocf_rb_tree_remove(&seq_cutoff->tree, node);
+			ocf_rb_tree_insert(&seq_cutoff->tree, node);
 		}
-		list_move_tail(&stream->list, &core->seq_cutoff->lru);
-	} else {
-		stream = list_first_entry(&core->seq_cutoff->lru,
+		list_move_tail(&stream->list, &seq_cutoff->lru);
+
+		return stream;
+	}
+
+	if (insert) {
+		stream = list_first_entry(&seq_cutoff->lru,
 				struct ocf_seq_cutoff_stream, list);
-		ocf_rb_tree_remove(&core->seq_cutoff->tree, &stream->node);
-		stream->rw = req->rw;
-		stream->last = req->byte_position + req->byte_length;
-		stream->bytes = req->byte_length;
-		ocf_rb_tree_insert(&core->seq_cutoff->tree, &stream->node);
-		list_move_tail(&stream->list, &core->seq_cutoff->lru);
+		ocf_rb_tree_remove(&seq_cutoff->tree, &stream->node);
+		stream->rw = rw;
+		stream->last = addr + len;
+		stream->bytes = len;
+		stream->req_count = 0;
+		stream->valid = true;
+		ocf_rb_tree_insert(&seq_cutoff->tree, &stream->node);
+		list_move_tail(&stream->list, &seq_cutoff->lru);
+
+		return stream;
+	}
+
+	return NULL;
+}
+
+void ocf_core_seq_cutoff_base_promote(struct ocf_seq_cutoff *seq_cutoff,
+		struct ocf_seq_cutoff_stream *src_stream)
+{
+	struct ocf_seq_cutoff_stream *dst_stream;
+
+	dst_stream = list_first_entry(&seq_cutoff->lru,
+			struct ocf_seq_cutoff_stream, list);
+	ocf_rb_tree_remove(&seq_cutoff->tree, &dst_stream->node);
+	dst_stream->rw = src_stream->rw;
+	dst_stream->last = src_stream->last;
+	dst_stream->bytes = src_stream->bytes;
+	dst_stream->req_count = src_stream->req_count;
+	dst_stream->valid = true;
+	ocf_rb_tree_insert(&seq_cutoff->tree, &dst_stream->node);
+	list_move_tail(&dst_stream->list, &seq_cutoff->lru);
+	src_stream->valid = false;
+}
+
+void ocf_core_seq_cutoff_update(ocf_core_t core, struct ocf_request *req)
+{
+	ocf_seq_cutoff_policy policy = ocf_core_get_seq_cutoff_policy(core);
+	uint32_t promotion_threshold =
+			ocf_core_get_seq_cutoff_promotion_threshold(core);
+	uint32_t promotion_count =
+			ocf_core_get_seq_cutoff_promotion_count(core);
+	struct ocf_seq_cutoff_stream *stream;
+	bool promote = false;
+
+	if (policy == ocf_seq_cutoff_policy_never)
+		return;
+
+	if (req->byte_length >= promotion_threshold)
+		promote = true;
+
+	if (promotion_count == 0)
+		promote = true;
+
+	if (req->seq_cutoff_core || promote) {
+		env_rwlock_write_lock(&core->seq_cutoff->lock);
+		stream = ocf_core_seq_cutoff_base_update(core->seq_cutoff,
+				req->byte_position, req->byte_length, req->rw,
+				promote);
+		env_rwlock_write_unlock(&core->seq_cutoff->lock);
+
+		if (!stream) {
+			ocf_core_seq_cutoff_base_update(
+					req->io_queue->seq_cutoff,
+					req->byte_position, req->byte_length,
+					req->rw, true);
+		}
```

Comment:
At this point queue stream might be due to promotion. If you replace lines 317-324 with simple "if (stream) return" this would be covered.

Reply:
Nope. If `promote` is `true` then line 312 should never fail.

Reply:
And if it isn't and there is no matching core stream, you update queue stream and don't consider the case that it might be time to promote it right away.

Reply:
This is never the case as sequential cutoff is handled for one queue at the time (there is no requeueing between check and update). However this may change in future and proposed modification makes implementation resistant to changes like that.